File: Java/13. Roman to Integer.java
class Solution {
    public int romanToInt(String s) {
        Map<Character, Integer> numbers = new HashMap<Character, Integer>() {
            {
                put('I', 1);
                put('V', 5);
                put('X', 10);
                put('L', 50);
                put('C', 100);
                put('D', 500);
                put('M', 1000);
            }            
        };
        int res = 0, flag = 1;
        for (int i = 0; i < s.length(); i++) {
            flag = 1;
            if (i < s.length() -1 && numbers.get(s.charAt(i)) < numbers.get(s.charAt(i + 1)))
                flag = -1;
            res += flag * numbers.get(s.charAt(i));
        }
        return res;
    }
}

File: 2024/13. Roman to Integer.cpp
class Solution {
public:
    unordered_map<char, int> ri = {{'M', 1000}, {'D', 500}, {'C', 100},
                                   {'L', 50},   {'X', 10},  {'V', 5},
                                   {'I', 1}};
    int romanToInt(string s) {
        int num = 0;
        for (int i = 0; i < s.length(); i++) {
            int flag = 1;
            if (i + 1 < s.length() &&
                ((s[i] == 'C' && (s[i + 1] == 'D' || s[i + 1] == 'M')) ||
                 (s[i] == 'X' && (s[i + 1] == 'L' || s[i + 1] == 'C')) ||
                 (s[i] == 'I' && (s[i + 1] == 'V' || s[i + 1] == 'X')))) {
                flag = -1;
            }
            num += flag * ri[s[i]];
        }
        return num;
    }
};